cpp/src/fil/fil.cu

```diff
@@ -325,11 +320,33 @@ struct forest {
   float global_bias_ = 0;
   shmem_size_params class_ssp_, proba_ssp_;
   int fixed_block_count_ = 0;
+  int max_shm_           = 0;
   // Optionally used
   rmm::device_uvector<float> vector_leaf_;
   cat_sets_device_owner cat_sets_;
 };
 
+template <typename storage_type>
+struct enable_smem_carveout : dispatch_functor<void> {
```

Comment:
`enable_smem_optin` would be a better name, but this is up to you.

Reply:
tried `opt_into_arch_dependent_smem`